test/e2e/targets/azureeventhubs/main.go

```diff
@@ -190,12 +191,105 @@ var _ = Describe("Azure Event Hubs target", func() {
 
 				By("verifying the sent event", func() {
 					Expect(len(payload)).To(BeNumerically(">", 0))
+					var ce cloudevents.Event
 
-					// NOTE: The payload will be a stringified version of the CloudEvent
-					Expect(payload).To(ContainSubstring(string(event.Data())))
-					Expect(payload).To(ContainSubstring("type: " + event.Type()))
-					Expect(payload).To(ContainSubstring("source: " + event.Source()))
-					Expect(payload).To(ContainSubstring("id: " + event.ID()))
+					err := json.Unmarshal(payload, &ce)
+					Expect(err).ToNot(HaveOccurred())
+
+					Expect(ce.Data()).To(ContainSubstring(string(event.Data())))
+					Expect(ce.ID()).To(Equal(event.ID()))
+					Expect(ce.Type()).To(Equal(event.Type()))
+					Expect(ce.Subject()).To(Equal(event.Subject()))
```

Comment:
Subject isn't a mandatory context attribute, but `Source` is. We can keep the assertion of Subject, but Source shouldn't be missing.

https://github.com/triggermesh/triggermesh/blob/12b2f1e86ba081390d82f1db852059777a0e1159/test/e2e/targets/awss3/main.go#L168

It would also be nice to add an additional assertion to ensure that the rogue CE extension is preserved during serialization:

https://github.com/triggermesh/triggermesh/blob/12b2f1e86ba081390d82f1db852059777a0e1159/test/e2e/targets/awss3/main.go#L170-L171

Reply:
Am aware Source is not mandatory, but it is being set by the hello event. Also good catch on the extension.

Reply:
What I mean is that, because the Hello event contains a Source attribute, this attribute should also be asserted.

Rationale: ensure that this field isn't being transformed by the target (intentionally or not), since the contract is that the target preserves all CE attributes.

Reply:
Okay now I see what you mean. Added.